internal/backends/sqlite/collection.go

```diff
@@ -32,304 +30,146 @@ import (
 
 // collection implements backends.Collection interface.
 type collection struct {
-	db   *database
-	name string
+	r      *metadata.Registry
+	dbName string
+	name   string
 }
 
 // newDatabase creates a new Collection.
-func newCollection(db *database, name string) backends.Collection {
+func newCollection(r *metadata.Registry, dbName, name string) backends.Collection {
 	return backends.CollectionContract(&collection{
-		db:   db,
-		name: name,
+		r:      r,
+		dbName: dbName,
+		name:   name,
 	})
 }
 
 // Query implements backends.Collection interface.
 func (c *collection) Query(ctx context.Context, params *backends.QueryParams) (*backends.QueryResult, error) {
-	conn, err := c.db.b.pool.DB(c.db.name)
-	if err != nil {
-		return nil, err
+	db := c.r.DatabaseGetExisting(ctx, c.dbName)
+	if db == nil {
+		return &backends.QueryResult{
+			Iter: newQueryIterator(ctx, nil),
+		}, nil
 	}
 
-	table, err := c.db.b.metadataStorage.tableName(ctx, c.db.name, c.name)
-	if err != nil {
-		return nil, err
-	}
+	tableName := c.r.CollectionToTable(c.name)
 
-	query := fmt.Sprintf(`SELECT sjson FROM "%s"`, table)
+	query := fmt.Sprintf(`SELECT _ferretdb_sjson FROM %q`, tableName)
 
-	rows, err := conn.QueryContext(ctx, query)
+	rows, err := db.QueryContext(ctx, query)
 	if err != nil {
-		return nil, err
+		return nil, lazyerrors.Error(err)
 	}
 
 	return &backends.QueryResult{
-		DocsIterator: newQueryIterator(ctx, rows, &queryIteratorParams{unmarshal: sjson.Unmarshal}),
+		Iter: newQueryIterator(ctx, rows),
 	}, nil
 }
 
 // Insert implements backends.Collection interface.
 func (c *collection) Insert(ctx context.Context, params *backends.InsertParams) (*backends.InsertResult, error) {
-	err := c.db.CreateCollection(ctx, &backends.CreateCollectionParams{Name: c.name})
-	if err != nil {
-		return nil, err
-	}
-
-	conn, err := c.db.b.pool.DB(c.db.name)
-	if err != nil {
-		return nil, err
-	}
-
-	table, err := c.db.b.metadataStorage.tableName(ctx, c.db.name, c.name)
-	if err != nil {
-		return nil, err
-	}
-
-	tx, err := conn.BeginTx(ctx, nil)
-	if err != nil {
-		return nil, err
+	if _, err := c.r.CollectionCreate(ctx, c.dbName, c.name); err != nil {
+		return nil, lazyerrors.Error(err)
 	}
 
-	defer func() {
-		if err != nil {
-			// TODO: check error
-			tx.Rollback()
-		}
-	}()
+	// TODO https://github.com/FerretDB/FerretDB/issues/2750
 
-	var inserted int64
+	db := c.r.DatabaseGetExisting(ctx, c.dbName)
+	query := fmt.Sprintf(`INSERT INTO %q (_ferretdb_sjson) VALUES (?)`, c.r.CollectionToTable(c.name))
```

Comment:
We are changing the column name here, I believe sqlite is still experimental so backwards compatibility isn't an issue I suppose.

Reply:
Yeah, and now we use a reserved prefix so we could implement #226 more easily

Reply:
Yeah that's good direction :+1: